[ADD] estate: created a new module

pimai-odoo opened the pull request:
Created and configured the database for the Real Estate module, defining necessary fields such as property type, tags, buyer, salesperson, and offers. Implemented list, form, and search views with groupings, and added computed fields, inverse functions, and onchange functionalities. Included 'Sold' and 'Cancel' buttons with their corresponding functionality, as well as 'Accept' and 'Refuse' buttons with related logic.

pimai-odoo committed:
[ADD] estate: created a new module

Added a new 'real estate' module, created the database, and defined necessary
columns.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1013.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1013)

bit-odoo commented on estate/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import models
```
There is a missing "blank" line at the end of the file.

bit-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import estate_property
```
There is a missing "blank" line at the end of the file.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+        selection=[('north', 'North'),('south', 'South'),('east', 'East'),('west', 'West')],
+        string="Garden Orientation"
+    )
+    active = fields.Boolean(string="Active", default=True)
```
There is a missing "blank" line at the end of the file.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+    _name = "estate.property"
+    _description = "Real Estate Property"
+
+    name = fields.Char(required='True')
```
It will throw an error. It takes a boolean value. Test your code before pushing it to GitHub.
```suggestion
    name = fields.Char(required=True)
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,23 @@
+from odoo import models,fields
+
+class estateproperty(models.Model):
```
Leave two lines between import and classes for better readability.

pimai-odoo committed:
[IMP] estate: update fields and menus in real estate

Set default values for fields, added new field attributes, and created menus
for accessing default list and form views in Real Estate module

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,8 @@
+    'depends': ['base'],
+    'data': ['security/ir.model.access.csv',
+    'views/estate_property_views.xml',
+    'views/estate_menus.xml'
```
```suggestion
    'data': [
        'security/ir.model.access.csv',
        'views/estate_property_views.xml',
        'views/estate_menus.xml'
```

bit-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+estate.access_estate_property,access_estate_property,estate.model_estate_property,base.group_user,1,1,1,1
```
There is a missing "blank" line at the end of the file.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,7 @@
+            <menuitem id="menu_estate_property_action" action="action_estate_property" />
+        </menuitem>
+    </menuitem>
+</odoo>
```
There is a missing "blank" line at the end of the file.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,12 @@
+            </p>
+        </field>
+    </record>
+</odoo>
```
There is a missing "blank" line at the end of the file.

bit-odoo reviewed (commented):
Hello,
Nice work!
I have added some comments please have a look at it.

module name is `estate` not `real_estate`.you can addapt it in commit message title.
For commit message you can refer this documentation :- https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html#commit-message-full-description

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,42 @@
+        copy=False,
+        default="new",
+    )
+    active = fields.Boolean(string="Active", default=True)
```
By default odoo generate string. If your string is same as technical name, then no need to add.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,42 @@
+from odoo import models, fields
+from dateutil.relativedelta import relativedelta
+
+class estateproperty(models.Model):
```
Leave two lines between import and classes for better readability.

bit-odoo reviewed (commented):
I have added some comments.

pimai-odoo committed:
[IMP] estate: add list/form/search views with group by in Real Estate

Created list, form, and search views, and added group by functionality in the Real Estate.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+from dateutil.relativedelta import relativedelta
+
+
+class estateproperty(models.Model):
```
Class name should be in camel case 'EstateProperty`.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+    facades = fields.Integer("Facades")
+    garage = fields.Boolean("Garage")
+    garden = fields.Boolean("Garden")
+    garden_area = fields.Integer("Garden Area (sqm)")
```
Please adapt these changes that were suggested to you earlier. You don't have to write a string if it's the same as the technical name.
https://github.com/odoo/tutorials/pull/1013#discussion_r2471806379

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+    )
+    active = fields.Boolean(string="Active", default=True)
+    
+
```
One empty line is sufficient.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+    name = fields.Char(required=True)
+    description = fields.Text("Description")
+    postcode = fields.Char("Postcode")
+    date_availability = fields.Date("Availability Date", default=fields.Date.today()+relativedelta(months=3))
```
```suggestion
    date_availability = fields.Date("Availability Date", default=fields.Date.today() + relativedelta(months=3))
```

bit-odoo reviewed (commented):
Hello, nice work!

Please use this documentation:- https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html#commit-message-full-description

do not push commit like "WIP", and make sure that your runbot is green.

pimai-odoo committed:
[IMP] estate: added property type,tags,buyer,salesperson & offers in Real Estate

Added Property Type and Tag models, buyer and salesperson fields,
and the Property Offer model in the Real Estate module.

pimai-odoo committed:
[IMP] estate: add computed fields,inverse funcs, onchange & Sold/Cancel buttons

Added computed fields with inverse functions and onchange methods,
along with Sold and Cancel buttons to enhance the
Real Estate module functionality.

bit-odoo commented:
Hello!
can you please make sure that your runbot is green.
https://runbot.odoo.com/runbot/tutorials-12?search=-pimai

And please update the PR message.

pimai-odoo committed:
[IMP] estate: added Accept,Refuse buttons & related functionality in Real Estate

Added Accept and Refuse buttons along with the supporting functionality
to improve workflow in the Real Estate module.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,88 @@
+from odoo import models, fields, api
+from dateutil.relativedelta import relativedelta
+from odoo.exceptions import UserError
```
```suggestion
from dateutil.relativedelta import relativedelta

from odoo import models, fields, api
from odoo.exceptions import UserError
```
First external libraries and then internal import.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,88 @@
+            if record.state == 'sold':
+                raise UserError("A sold property cannot be cancelled")
+            else:
+                self.state = 'cancelled'
```
It will throw an error if multiple records are in self.

bit-odoo replied:
What about these changes?

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,88 @@
+            if record.state == 'cancelled':
+                raise UserError("A cancelled property cannot be set as sold")
+            else:
+                self.state = 'sold'
```
Same here.

bit-odoo replied:
here also

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,45 @@
+from odoo import fields, models, api
+from dateutil.relativedelta import relativedelta
```
same as the previous file.

bit-odoo reviewed (commented):
Hello,
I have added some remarks :)

pimai-odoo committed:
[IMP] estate: add SQL & Python constraints for price & uniqueness in Real Estate

Added SQL constraints to ensure positive expected, selling,
and offer prices, and unique names for property tags and types.
Added a Python constraint to ensure the selling price
is not less than 90% of the expected price.

pimai-odoo committed:
[IMP] estate: added Inline Views, Widgets, List Order, Attributes, & Options

Implemented Inline Views and Widgets to enhance the user experience in
Real Estate. Added List Order functionality for improved organization and
sorting of listings. Added Attributes and Options to increase customization
and flexibility for users.

pimai-odoo committed:
[IMP] estate: added stat button, related fields & business logic updates

Added a stat button to the Property Type model and linked offers through
related fields. Implemented offer count computation and display on button click.
Updated business logic to restrict property deletion unless state is
‘New’ or ‘Cancelled’, set property state to ‘Offer Received’ on offer creation,
and prevent creation of lower offers than existing ones.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,65 @@
+            if property_id and offer_price is not None:
+                estate_property = self.env['estate.property'].browse(property_id)
+                estate_property.state = 'offer_received'
+                existing_offers = self.search([('property_id', '=', property_id), ('price', '>=', offer_price)])
```
It is not good way to use search() inside a loop.

bit-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,23 @@
+
+    name = fields.Char(required=True)
+    property_ids = fields.One2many("estate.property", "property_type_id", string="Properties")
+    sequence = fields.Integer("Sequence", default=1)
```
As I suggested to you in previous reviews, Odoo always generates a string if it same as the technical name, then no need to add.

bit-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,23 @@
+    property_ids = fields.One2many("estate.property", "property_type_id", string="Properties")
+    sequence = fields.Integer("Sequence", default=1)
+    offer_ids = fields.One2many("estate.property.offer", "property_type_id", string="Offers")
+    offer_count = fields.Integer(string="Offer Count", compute="_compute_offer_count")
```
Same here.

pimai-odoo committed:
[IMP] estate: added user field, new module and start invoice process

Added a new field in the user form to show linked properties.
Created a new module to connect Real Estate with accounting features.
Began the invoice creation process when a property is marked as sold.

pimai-odoo committed:
[IMP] estate: added invoice lines for sold properties and Kanban view

Added two invoice lines during invoice creation for sold properties:
one for 6% of the selling price and another for administrative fees of 100.00.
Added a Kanban view to improve property visualization in the Real Estate module.
